crates/wasi-http/src/types_impl.rs

```diff
@@ -258,7 +294,17 @@ impl<T: WasiHttpView> crate::bindings::http::types::HostOutgoingRequest for T {
         authority: Option<String>,
         headers: Resource<Headers>,
     ) -> wasmtime::Result<Resource<HostOutgoingRequest>> {
-        let headers = get_fields_mut(self.table(), &headers)?.clone();
+        let (ty, fields) = get_fields_mut(self.table(), &headers)?;
+
+        if !matches!(ty, FieldsType::Request) {
+            return Err(Error::FieldsError(
+                "Response fields used when constructing a Request".to_owned(),
+            )
+            .into());
+        }
```

Comment:
I don't like this trap when the fields provided aren't meant for a request. This pushes me towards thinking that we shouldn't use a `constructor` here, but instead a static method for making a request that can fail.